src/System.Management.Automation/engine/Utils.cs

```diff
@@ -33,6 +34,223 @@ namespace System.Management.Automation
+            }
+
+            i = (int)value;
+            return true;
+        }
+
+        internal static bool TryCast(BigInteger value, out uint u)
+        {
+            if (value < uint.MinValue || uint.MaxValue < value)
+            {
+                u = 0;
+                return false;
+            }
+
+            u = (uint)value;
+            return true;
+        }
+
+        internal static bool TryCast(BigInteger value, out long l)
+        {
+            if (value < long.MinValue || long.MaxValue < value)
+            {
+                l = 0;
+                return false;
+            }
+
+            l = (long)value;
+            return true;
+        }
+
+        internal static bool TryCast(BigInteger value, out ulong ul)
+        {
+            if (value < ulong.MinValue || ulong.MaxValue < value)
+            {
+                ul = 0;
+                return false;
+            }
+
+            ul = (ulong)value;
+            return true;
+        }
+
+        internal static bool TryCast(BigInteger value, out decimal dm)
+        {
+            if (value < (BigInteger)decimal.MinValue || (BigInteger)decimal.MaxValue < value)
+            {
+                dm = 0;
+                return false;
+            }
+
+            dm = (decimal)value;
+            return true;
+        }
+
+        internal static bool TryCast(BigInteger value, out double db)
+        {
+            if (value < (BigInteger)decimal.MinValue || (BigInteger)decimal.MaxValue < value)
+            {
+                db = 0;
+                return false;
+            }
+
+            db = (double)value;
+            return true;
+        }
+
+        /// <summary>
+        /// Parses a given string or ReadOnlySpan&lt;char&gt; to calculate its value as a binary number.
+        /// Assumes input has already been sanitized and only contains zeroes (0) or ones (1).
+        /// </summary>
+        /// <param name="digits">Span or string of binary digits. Assumes all digits are either 1 or 0.</param>
+        /// <param name="unsigned">
+        /// Whether to treat the number as unsigned. When false, respects established conventions
+        /// with sign bits for certain input string lengths.
+        /// </param>
+        /// <returns>Returns the value of the binary string as a BigInteger.</returns>
+        internal static BigInteger ParseBinary(ReadOnlySpan<char> digits, bool unsigned)
+        {
+            if (!unsigned)
+            {
+                if (digits[0] == '0')
+                {
+                    unsigned = true;
+                }
+                else
+                {
+                    switch (digits.Length)
+                    {
+                        // Only accept sign bits at these lengths:
+                        case 8: // byte
+                        case 16: // short
+                        case 32: // int
+                        case 64: // long
+                        case 96: // decimal
```

Comment:
To my knowledge, there aren't any supported platforms where these numbers will be different. However, if there is a programmatic way to get expected sign bit lengths, I would be more than open to checking against this instead.

Reply:
@vexx32 It is not low level problem because we say about `digits` - it can be PowerShell language propblem if ARM user expect sign bit on another position.

Reply:
Aye, that could definitely be a concern. I could always refactor this to incorporate this, but I'd need to know what to incorporate here to account for it. Not sure where to find this information at the moment. 🙁

Reply:
I found nothing useful. I hope @mklement0 help.

Reply:
I'll take a look at it tomorrow, @iSazonov, but I fear I won't be of much help.

Reply:
As for CPU architectures: I don't think we need to worry about those, as long as the behavior is well-documented in terms of what bit patterns result in what .NET primitive types.

Note that we already abstract endianness away by always interpreting a byte-spanning bit / hex-digit pattern from left to right (significant bits first).

However, note that this proposed implementation does _not_ match the existing parsing of _hex_ constants:

```powershell
PS> 0xff; 0b11111111; 0b011111111 
255 # unsigned
-1  # !! signed
255 # unsigned again, simply due to prepending 0
```

We should at least remain consistent with hex parsing, which - based on my experiments (haven't looked at the source) is not based on fixed bit positions - leading zeros make no difference -  but is simply based on the smallest _signed_ integer type >= `[int32]` that can accommodate the value as a _bit pattern_, meaning irrespective of its sign, which currently (before this PR) tops out at `[int64]`:

```powershell
PS> 0xffffffff; 0x0ffffffff  # 32 effective bits
-1  # !! bit pattern fits into [int32], and because the high bit happens to be set, the result is *negative*
-1  # behavior is not dependent on specific number of digits

PS> 0xffffffffffffffff; 0x0ffffffffffffffff # 64 effective bits
-1  # !! bit pattern fits into [int64], and because the high bit happens to be set, the result is *negative*
-1  # behavior is not dependent on specific number of digits
```

Therefore:

* leading zeros should be _ignored_.

* the boundaries should be just `32` and `64` for consistency with current hex-literal parsing, and `96` for the new widening to `[decimal]`.

* parsing of _decimal_ literals > `[decimal]` is seemingly _broken_ by this PR; e.g., the equivalent of `0x1000000000000000000000000` (`[decimal]::MaxValue + 1`) is decimal `79228162514264337593543950336`, which in v6.1.0 yields `7.92281625142643E+28`, i.e. a `[double]`(!)

```powershell
PS>  79228162514264337593543950336 # in 6.1.0: 7.92281625142643E+28
...
The numeric constant 79228162514264337593543950336 is not valid. # !! ERROR, in this PR
...
```

Note that there's a current inconsistency in that _hex_ literals do not widen beyond `[int64]`, whereas _decimal_ ones do, as shown above, but ultimately to _`[double]`_ (-> `[decimal]` -> `[double`)

While ultimately widening to _`[bigint]`_ instead of `[double]` makes sense for _decimal_ literals too, it would technically be a breaking change we should discuss; at the very least the existing behavior with decimal literals shouldn't be broken.

Reply:
Thank you very much for taking the time to set this out! 😄 

So... yes, I agree. *Mostly*.

The current hexadecimal parsing behaviour has been maintained as-is for compatibility reasons; I don't want folks who've been using `0xFF` for `255` to suddenly find their literals coming back as `-1` and so forth.

*However*, given that binary strings are typically really most useful at the byte level for dealing with and helping to conceptualize bitwise operations, I don't think it's at all useful to copy the present hex behaviour of signed bits. *Very few* people are going to want to write out 32 digits just to get a sign bit. And, yes, you *can* use a negative-sign prefix to bypass that and just flip your numbers. However, given that when working with binary literals users will largely want them to mimic expected bit-pattern behaviours and that binary strings will be unwieldy to use at int32/64 lengths and beyond, I cannot imagine that it is meaningful to restrict sign bits to 32, 64, and 96 digit lengths. 

After all, if they *need* the numeral to be unsigned, they need only append the `u` suffix. (Same for hex, incidentally, but I doubt that will be changing anytime soon.)

Really, the only reason hex literals don't widen beyond int64 is simply due to the fact that there is no native `TryParse()` method that allows it for `double` and other types, which is rendered irrelevant with the methods I chose to use here.

Regarding your point about leading zeroes... currently they are ignored, yes, and I am unsure whether that ought to be retained? Given they are currently ignored, I'm not sure I'd expect many scripts to bother including leading zeros, but I'll leave that choice up to folks cleverer than I.

Let me check that decimal parsing case... I might have let a code path slip me by, it's no matter. Just gotta make sure the integer widens to double as it needs to. I'll have that fixed soon -- nice catch! 😄

Reply:
@mklement0 yep, I made an error in the TryCast overload for double casting. Thanks for catching that!

Reply:
My pleasure; thank you for all the work on typed number literals - great additions to the language - and your thoughtful contributions to this community in general.

---

Generally, yes, the use of `u` avoids the whole sign-bit problem - but you have to remember to use it.

_By default_, I would think that you would _usually_ want _positive_, i.e., _unsigned_ numbers, and I wish that's how it had always worked with hex. literals - sadly, that ship has sailed.

Given that binary / hex literals are typically used for bit fields / flag enumeration values, there's rarely a reason for _negative_ numbers - or am I missing important use cases?

Personally, I don't think of binary literals as _byte_-oriented, because 16-bit bit fields aren't unusual, for instance, though in general, you'll be dealing with _smaller_ numbers than in hex. notation, I presume.

Tying the signed-bit use to the _resulting type_, which is how hex. literals currently do it, is a simple rule that could apply to binary literals equally.

In fact, from a bit-field perspective, the **current implementation yields unexpected and ambiguous results**, because the _result type_'s `-1` bit pattern is substituted for the input's:

E.g., `0b11111111` is interpreted as a signed _byte_ and therefore becomes `-1`, but it doesn't become `-1` as a (signed) _byte_, but as the `Int32` instance it is converted to (as the smallest return type chosen), whose bit pattern is then obviously different: `0b11111111111111111111111111111111`; to put it differently:   
currently, `0b11111111`, `0b1111111111111111`, and `0b11111111111111111111111111111111` all turn in the _very same number_: `-1` as an `[int]`.

I suppose the inferred result type could be made more granular for _binary_ literals, and start with `[sbyte]`, then widen to `[short]`, but I'm not sure that's the way to go, both for reasons of diverging from hex.-literal behavior and for the questionable utility of the interpretation as a signed type in general.

Staying consistent with hex. literals additionally has the up-side that you're less likely to run into the potentially surprising sign-bit behavior, given that binary literals will typically be smaller in value; i.e., you're less likely to get a negative number when you didn't expect one.

---

Re leading zeros: Yes, it would be odd for users to include them to begin with.

I think simply ignoring them is an easy and easily documentable solution to this problem, especially given that they currently (pre-PR) serve no useful purpose in hex. literals - they do not widen the type. 

And neither do they widen the type of the binary literal type in this PR; all they do is cause the - obscure - change in signed/unsigned behavior.

Reply:
Yeah, I hear you. I actually *did* experiment with having binary (and hex as well) return sub-int32 types, all the way down to byte and sbyte.

Given that in a majority of cases any actual arithmetic automatically upscales them appropriately in any case, this isn't a *terrible* solution. Actually, due to its backing in C#, any arithmetic performed on lower types automatically converts values up to int32 in many cases. However, it *would* be a bit of an oddity and ultimately I agree that it would probably do more harm than good.

I suppose that given the fact that one *can* ultimately flip the value with a prefixed negative sign, it's much of a muchness which sign bit lengths the parser respects, as it really is entirely up to us to figure out the most effective and useful behaviour in this instance.

Is it worth looking at how C# implements binary literals and attempting to match that relatively closely, do you think?

Reply:
C# exhibits consistent behavior across hex and binary literals, and implicitly promotes from signed to unsigned types as needed; in other words: you always get _positive_ numbers - it's how I wish it worked in PowerShell too.

As in PowerShell, the smallest type chosen is `Int32`.

```csharp
// Hex
0x1 // -> Int32
0x7fffffff // -> Int32 - int.MaxValue
0x80000000 // -> UInt32 - int.MaxValue + 1

// Binary equivalents
0b1 // -> Int32  
0b01111111111111111111111111111111 // -> Int32
0b10000000000000000000000000000000 // -> Uint32

```

Reply:
That would be nice to do, and well within the realm of possibility. I suppose *that* depends on whether in not it is considered an allowable change at this point in time :s

Reply:
Good question, @vexx32.

Perhaps, if we're lucky, the following changes  will fall into [Bucket 3: Unlikely Grey Area](https://github.com/PowerShell/PowerShell/blob/master/docs/dev-process/breaking-change-contract.md#bucket-3-unlikely-grey-area):

* Treat _all_ number bases in literals the _same_ - as C# does (decimal, hex, binary)
   
* Use `[int32]` as the smallest type, and promote _as needed_ from signed to unsigned, as shown above - resulting in _positive numbers always_.

     * This is a deviation from current hex.-literal parsing.

* (PowerShell-specific) For values > `[int64]::MaxValue` (which C# doesn't support), promote to `[decimal]` first, and, if > `[decimal]::MaxValue`, to `[bigint]`

     * This is a deviation from current decimal-literal parsing (which promotes to `[double]` for values > `[decimal]::MaxValue`

What do you think, @SteveL-MSFT?

Reply:
From docs https://docs.microsoft.com/en-us/dotnet/api/system.numerics.biginteger?view=netframework-4.7.2
> When parsing a hexadecimal string, the BigInteger.Parse(String, NumberStyles) and BigInteger.Parse(String, NumberStyles, IFormatProvider) methods assume that if the most significant bit of the first byte in the string is set, or if the first hexadecimal digit of the string represents the lower four bits of a byte value, the value is represented by using two's complement representation. For example, both "FF01" and "F01" represent the decimal value -255. To differentiate positive from negative values, positive values should include a leading zero. The relevant overloads of the ToString method, when they are passed the "X" format string, add a leading zero to the returned hexadecimal string for positive values. This makes it possible to round-trip BigInteger values by using the ToString and Parse methods, as the following example shows.

Reply:
Aye, that's the BigInteger parse method. C#'s own parsing rules for hex / binary literals in code is very different. I understand *why* there are so many varying ways that these can be parsed in C#, but it's a bit baffling to me that they don't implement any kind of direct standard parsing for hex/binary that *does* mimic their native parsing. Very odd!

Reply:
Good points, @vexx32.

@iSazonov, remember that `[bigint]` here is just a _helper type_ behind the scenes for implementing arbitrarily large binary and hex _source-code_ literals (and possible decimal ones, based on the suggestions above) - we are _not_ trying to emulate `[bigint]`'s runtime parsing of strings.

As shown above, C# - sensibly - interprets _all_ hex and binary source-code literals as _positive_ numbers, choosing the smallest signed or unsigned type >= `Int32` to accommodate that number, up to `UInt64`.

Regrettably, PowerShell doesn't currently do that for hex numbers, hence my Bucket 3 suggestion above.

If we get the green light to fix this, it means applying the same logic to the > `UInt64` ranges we support:

 * <= `0x7fffffffffffffffffffffff` (and its binary equivalent) -> `[decimal]`
*  &gt;= `0x800000000000000000000000` (and its binary equivalent) -> `[bigint]`

Reply:
@vexx32: Another thing about the current PR: a hex literal with >= 40 `f` digits reverts to an `Int32`:

```powershell
PS> (0xffffffffffffffffffffffffffffffffffffffff).GetType().FullName
System.Int32
```

While the seemingly equivalent binary literal composed of 160 `1` digits actually remains a `[bigint]`, it still turns into a negative number:

```powershell
PS> 0b1111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111
-1
```

Reply:
Yep! I'm already working around how BigInt *wants* to parse things and special-casing things to mimic present behaviour. It would only simplify the special casing that I'd need to perform if we can opt to emulate C#'s interpretation of binary/hexadecimal literals.

Reply:
@vexx32: The always-positive logic is fairly easy to implement, if I understand things correctly:

From a hex string: prepend a `0`:

```csharp
// 0xff as 255; without the leading 0, the result would be -1
BigInteger.Parse("0ff",NumberStyles.AllowHexSpecifier)
```

From a byte array - prepend a `0x` byte (with Big-Endian ordering):

```csharp
new BigInteger(new byte[] { 0x0, 0xff }, isBigEndian: true)
```

Reply:
Even easier, actually. Hex strings are currently given a leading zero by default with my current code. I am then stripping it out in cases where it appears appropriate to allow a sign bit. I would simply remove the additional logic.

And with binary, it would simply always mean invoking the BigInteger constructor with a value of `true` for the `isUnsigned` parameter.

Reply:
Thinking further, I could even trim out the extra leading 0 if we eventually opt for parsing as unsigned, as it will be possible to simply instruct the BigInteger constructor to always read it as unsigned.

Reply:
Sorry for the delay, on the topic of the breaking change.  I know it's desirable to change it, but it seems a bit risky to make such a change.

Reply:
I figured that would probably be the case. It is a fairly straightforward follow up if we find reason to change it later.

I could even put it in as an experimental feature later on, I think.